lib/pengine/unpack.c

```diff
@@ -908,6 +908,82 @@ unpack_resources(const xmlNode *xml_resources, pcmk_scheduler_t *scheduler)
     return TRUE;
 }
 
+/*!
+ * \internal
+ * \brief Unpack essential information from topology request XML
+ *
+ * \param[in]  xml     Request XML to search
+ * \param[out] id      If not NULL, where to store level number
+ *
+ * \return Topology level XML from within \p xml, or NULL if not found
+ * \note The caller is responsible for freeing \p *target and \p *desc if set.
+ */
+static xmlNode *
+unpack_level_request(xmlNode *xml, int *id)
+{
+    int local_id = 0;
+
+    /* The level element can be the top element or lower. If top level, don't
+     * search by xpath, because it might give multiple hits if the XML is the
+     * CIB.
+     */
+    if ((xml != NULL) && !pcmk__xe_is(xml, PCMK_XE_FENCING_LEVEL)) {
+        xml = get_xpath_object("//" PCMK_XE_FENCING_LEVEL, xml, LOG_WARNING);
+    }
+
+    if (xml != NULL) {
+        crm_element_value_int(xml, PCMK_XA_INDEX, &local_id);
+    }
+
+    if (id != NULL) {
+        *id = local_id;
+    }
+
+    return xml;
+}
+
+/*!
+ * \internal
+ * \brief Parse configuration XML for fencing topology information
+ *
+ * \param[in]     xml_fencing_topology  Top of fencing topology configuration XML
+ * \param[in,out] scheduler             Scheduler data
+ *
+ * \return TRUE
+ */
+gboolean
+unpack_fencing_topology(const xmlNode *xml_fencing_topology, pcmk_scheduler_t *scheduler)
+{
+    xmlNode *xml_obj = NULL;
+    int id = 0;
+    xmlNode *level;
+
+    for (xml_obj = pcmk__xe_first_child(xml_fencing_topology, NULL, NULL, NULL);
+         xml_obj != NULL; xml_obj = pcmk__xe_next(xml_obj)) {
```

Comment:
Here, we can pass `PCMK_XE_FENCING_LEVEL` as the second argument to `pcmk__xe_first_child()`, and use `pcmk__xe_next_same()` instead of `pcmk__xe_next()`. That will ensure that every iteration is a `PCMK_XE_FENCING_LEVEL` child (if validation is enabled, the only other thing that would be allowed here is an XML comment). We don't need `level` since `xml_obj` will always be what we need.